xarray/tests/test_dataarray.py

```diff
@@ -1634,6 +1658,31 @@ def test_assign_coords_no_default_index(self) -> None:
         assert_identical(actual.coords, coords, check_default_indexes=False)
         assert "y" not in actual.xindexes
 
+    def test_assign_coords_extra_dim_index_coord(self) -> None:
+        class AnyIndex(Index):
+            # This test only requires that the coordinates to assign have an
+            # index, whatever its type.
+            pass
+
+        idx = AnyIndex()
+        coords = Coordinates(
+            coords={
+                "x": ("x", [1, 2]),
+                "x_bounds": (("x", "x_bnds"), [(0.5, 1.5), (1.5, 2.5)]),
+            },
+            indexes={"x": idx, "x_bounds": idx},
+        )
+
+        da = DataArray([1.0, 2.0], dims="x")
+        actual = da.assign_coords(coords)
+
+        # cannot use `assert_identical()` test utility function here yet
+        # (indexes invariant check is still based on IndexVariable, which
+        # doesn't work with AnyIndex coordinate variables here)
```

Comment:
Is the invariant test easy to update? It would be really nice to keep it up to date.

Reply:
Done in [bc578d8](https://github.com/pydata/xarray/pull/10116/commits/bc578d846bc9a6f38644c49091ea39320144920a) (this was actually already update for the Dataset invariant checks).